kuma/notifications/extract.py

```diff
@@ -0,0 +1,113 @@
+import functools
+import itertools
+import json
+import sys
+from collections.abc import Mapping
+
+bcd = json.loads(open("bcd.json").read())
+
+
+def walk(data, path=None, depth=sys.maxsize):
+    if depth == 0:
+        return
+
+    if not path:
+        path = ""
+
+    for key in data:
+        if not isinstance(data[key], Mapping):
+            continue
+
+        if key == "__compat":
+            # This is an element with compat information
+            yield path, data[key]
+        else:
+            yield from walk(
+                data[key], path=f"{path}{'.' if path else ''}{key}", depth=depth - 1
+            )
+
+
+def get_feature(bcd, feature):
+    try:
+        return functools.reduce(dict.get, feature.split("."), bcd) or {}
+    except (TypeError, KeyError):
+        return {}
+
+
+class NotificationGenerator:
+    text = "Notification for "
+
+    def __init__(self, path, old_bcd, new_bcd):
+        self.path = path
+        self.old_bcd = old_bcd
+        self.new_bcd = new_bcd
+        self.old_feature = get_feature(old_bcd, path).get("__compat", None)
+        self.new_feature = get_feature(new_bcd, path).get("__compat", None)
+
+    def support(self):
+        # Is this static?
+        old_support = self.old_feature.get("support", {}) if self.old_feature else {}
+        new_support = self.new_feature.get("support", {}) if self.new_feature else {}
+        browsers = set(list(old_support.keys()) + list(new_support.keys()))
+
+        for browser in browsers:
+            yield browser, old_support.get(browser, {}), new_support.get(browser, {})
+
+    def notification_text(self, template, browser):
+        return template.format(feature=self.path, browser=browser)
+
+    def compare(self, old, new):
+        raise NotImplemented
+
+    def generate_for_browsers(self):
+        for browser, old, new in self.support():
+            # ToDo: Review the spec and deal with lists
+            if isinstance(old, list):
+                old = old[0] if old else {}
+            if isinstance(new, list):
+                new = new[0] if new else {}
+
+            if template := self.compare(old, new):
+                yield self.notification_text(template, browser)
+
+    def generate(self):
+        notifications = list(self.generate_for_browsers())
+        if len(notifications) > 3:
+            return [self.text + "multiple browsers"]
+        elif len(notifications) > 0:
+            return [self.text + ", ".join(notifications)]
+        return [self.text + i for i in notifications]
+
+
+class NowSupported(NotificationGenerator):
+    text = "Now supported in "
+
+    def compare(self, old, new):
+        if new.get("version_added") and not old.get("version_added"):
```

Comment:
I want to alert you to something potentially hazardous here:

Setting aside lists (where a feature is represented by multiple support statements  as in `[{ "version_added": "X" }, …]`), it's also common for BCD to record data for future versions. That is, "now supported" is a question you can't answer from a support object alone. I suspect you'll need some sort of utility that looks up the current release of a browser (from the `browsers` part of the data), then uses that value to figure out whether a set of support statements shows support for that version.

The logic for this thing will be a bit complex: it'll have to be able to resolve versions in ranges (e.g., current version 92 in the range represented by `"version_added": "37", "version_removed": "99"`, multiple statements, and to ignore statements with flags or `"preview"` values).

Reply:
Thanks for the explanation! This is very useful! 

What I have right now is the naive implementation so I can run through the process of generating notifications for the whole bcd object.

Part of the reason I wanted to split the notification generators into classes is to be able to account for this kind of behaviour. I can now take what you wrote (plus anything else I can gather from the spec) and build a better NowSupported generator. I'll ping you when it's done so we can review it together

Reply:
Great, thanks @nicolaslara! Glad to hear that this is more of a placeholder than the real thing. I'm not surprised it was already on your mind, but I'm relieved to have it said explicitly.